src/compiler/types.ts

```diff
@@ -8531,29 +8531,75 @@ namespace ts {
-        <T extends Node>(nodes: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined;
+        <TIn extends Node | undefined, TVisited extends Node | undefined, TOut extends Node>(
+            node: TIn,
+            visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+            test: (node: Node) => node is TOut,
+            lift?: (node: readonly Node[]) => Node,
+        ): TOut | (TIn & undefined) | (TVisited & undefined);
+        <TIn extends Node | undefined, TVisited extends Node | undefined>(
+            node: TIn,
+            visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+            test?: (node: Node) => boolean,
+            lift?: (node: readonly Node[]) => Node,
+        ): Node | (TIn & undefined) | (TVisited & undefined);
     }
 
+    /**
+     * A function that walks a node array using the given visitor, returning an array whose contents satisfy the test.
+     *
+     * - If the input node array is undefined, the output is undefined.
+     * - If the visitor can return undefined, the node it visits in the array will be reused.
+     * - If the output node array is not undefined, then its contents will satisfy the test.
+     * - In order to obtain a return type that is more specific than `NodeArray<Node>`, a test
+     *   function _must_ be provided, and that function must be a type predicate.
+     *
+     * For the canonical implementation of this type, @see {visitNodes}.
+     */
     export interface NodesVisitor {
-        <T extends Node>(nodes: NodeArray<T>, visitor: Visitor | undefined, test?: (node: Node) => boolean, start?: number, count?: number): NodeArray<T>;
-        <T extends Node>(nodes: NodeArray<T> | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, start?: number, count?: number): NodeArray<T> | undefined;
-    }
-
-    export type VisitResult<T extends Node> = T | readonly T[] | undefined;
+        <TIn extends Node, TInArray extends NodeArray<TIn> | undefined, TVisited extends Node | undefined, TOut extends Node>(
+            nodes: TInArray,
+            visitor: Visitor<TIn, TVisited> | undefined,
+            test: (node: Node) => node is TOut,
+            start?: number,
+            count?: number,
+        ): NodeArray<TOut> | (TInArray & undefined);
+        <TIn extends Node, TInArray extends NodeArray<TIn> | undefined, TVisited extends Node | undefined>(
+            nodes: TInArray,
+            visitor: Visitor<TIn, TVisited> | undefined,
+            test?: (node: Node) => boolean,
+            start?: number,
+            count?: number,
+        ): NodeArray<Node> | (TInArray & undefined);
+    }
+
+    // Either a node, or a list of nodes to be be lifted via a lift function.
+    export type VisitResult<T extends Node | undefined> = T | readonly Node[];

```

Comment:
The diff sucks here but this used to be:

```
export type VisitResult<T extends Node> = T | readonly T[] | undefined;
```

So this is a breaking change; I need to ensure the type variable encodes undefined-ness so the NodeVisitor interface can use that to determine whether or not the output can be undefined. The array that the visitor returns doesn't need to have any type on it, because it's just getting passed to the lift function, which has signature `(nodes: readonly Node[]) => Node` (and gains nothing from knowing what its input type or output type actually are; the `test` does that).

Reply:
I've made this less breaky by making `Visitor`'s output be potentially undefined by default; only those who have explicitly used `VisitResult` have the potential to break. But, this is probably fine because if they don't return undefined, their visitor will be more accurate, and if they don't, they'll be forced to acknowledge that their visitor potentially returns undefined on certain paths and so the result needs to be checked (as was the case with many of our own visitors).